daemon/events/events.go

```diff
@@ -0,0 +1,144 @@
+package events
+
+import (
+	"context"
+	"time"
+
+	"github.com/alibaba/pouch/apis/types"
+
+	goevents "github.com/docker/go-events"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// Events is pubsub channel for events generated by the engine.
+type Events struct {
+	broadcaster *goevents.Broadcaster
+}
+
+// NewEvents return a new Events instance
+func NewEvents() *Events {
+	return &Events{
+		broadcaster: goevents.NewBroadcaster(),
+	}
+}
+
+// Publish sends an event. The caller will be considered the initial
+// publisher of the event. This means the timestamp will be calculated
+// at this point and this method may read from the calling context.
+// TODO: do we need lock when use broadcaster???
+func (e *Events) Publish(ctx context.Context, action string, eventType types.EventType, actor *types.EventsActor) error {
+	var err error
+	// ensure actor not nil
+	if actor == nil {
+		actor = &types.EventsActor{}
+	}
+
+	now := time.Now().UTC()
+	msg := types.EventsMessage{
+		Action:   action,
+		Type:     eventType,
+		Actor:    actor,
+		Time:     now.Unix(),
+		TimeNano: now.UnixNano(),
+	}
+
+	// compatibility with moby
+	switch eventType {
+	case types.EventTypeContainer:
+		msg.ID = actor.ID
+		msg.Status = action
+		if actor.Attributes != nil {
+			if image, ok := actor.Attributes["image"]; ok {
+				msg.From = image
+			}
+		}
+	case types.EventTypeImage:
+		msg.ID = actor.ID
+		msg.Status = action
+	}
+
+	defer func() {
```

Comment:
we don't defer here and can log the info if err = nil. defer is not straight forward